source/swap/contracts/Swap.sol

```diff
@@ -343,20 +350,17 @@ contract Swap is ISwap {
 
     // Ensure the transfer is not to self.
     require(from != to, "INVALID_SELF_TRANSFER");
-
-    if (kind == ERC721_INTERFACE_ID) {
-
-      // Attempt to transfer an ERC-721 token.
-      IERC721(token).transferFrom(from, to, param);
-
-    } else {
-      uint256 initialBalance = INRERC20(token).balanceOf(from);
-
-      // Attempt to transfer an ERC-20 token.
-      INRERC20(token).transferFrom(from, to, param);
-
-      // Ensure the amount has been transferred.
-      require(initialBalance.sub(param) == INRERC20(token).balanceOf(from), "TRANSFER_FAILED");
-    }
+    ITransferHandler transferHandler = registry.getTransferHandler(kind);
+    require(address(transferHandler) != address(0x0), "UNKNOWN_TRANSFER_HANDLER");
+
+    (bool success, bytes memory data) = address(transferHandler).
+      delegatecall(abi.encodeWithSelector(
```

Comment:
Why does this need to be a delegate call? Can this not just be implemented using `transferHandler.transferTokens(...)`?

Reply:
answered below.